File: src/csharp/1 - Two Sum.cs
class Solution
{
    public int[] TwoSum(int[] nums, int target)
    {
        return V2(nums, target);
    }

    private int[] V1(int[] nums, int target)
    {
        for (int i = 0; i < nums.Length; i++)
        {
            for (int j = i + 1; j < nums.Length; j++)
            {
                if (nums[i] + nums[j] == target)
                {
                    return new int[] { i, j };
                }
            }
        }
        throw new ArgumentException("no solution");
    }

    private int[] V2(int[] nums, int target)
    {
        var dict = new Dictionary<int, int>();
        for (int i = 0; i < nums.Length; i++)
        {
            if (dict.ContainsKey(target - nums[i]))
            {
                return new int[] { i, dict[target - nums[i]] };
            }
            // Add方法对于重复KEY会抛出异常，TryAdd会返回false。但这两者都没有更新目标键值对
            // 对于本题目，最终返回结果只需要保证是两个目标数字即可，具体是数组中的哪个无所谓，这里没有发生更新没问题
            dict.TryAdd(nums[i], i);
            // dict[nums[i]] = i;       // 添加或更新
        }
        throw new ArgumentException("no solution");
    }


    private static void Print(int[] nums, int target)
    {
        var s = new Solution();
        var res = s.TwoSum(nums, target);
        Console.WriteLine($"[{res[0]}, {res[1]}]");
    }

    static void Main()
    {
        Print(new int[] { 2, 7, 11, 15 }, 9);
        Print(new int[] { 3, 2, 4 }, 6);
        Print(new int[] { 3, 3 }, 6);
    }
}

File: src/fsharp/1 - Two Sum.fs
module Solution

open System.Collections.Generic

let V1 (array: int[]) target = 
    let mutable result = null
    for i = 0 to (array.Length - 1) do
        for j = 0 to (array.Length - 1) do
            if array.[i] + array.[j] = target then result <- [| i; j|]
    result

let V2 (array: int[]) target =
    let mutable result = null
    let mutable found = false
    let dict = new Dictionary<int, int>()
    for i = 0 to (array.Length - 1) do
        if not found then
            if dict.ContainsKey(target - array.[i]) then
                found <- true
                result <- [| i; dict.[target - array.[i]] |]
            else
                dict.TryAdd(array.[i], i) |> ignore
    result

// 更具F#自身特色的实现
let V3 (array: int[]) target = 
    let dict = array |> Seq.mapi (fun index num -> (num, index)) |> Map.ofSeq
    array
    |> Seq.mapi (fun index num -> (num, index))
    |> Seq.tryFind (fun (num, index) -> dict.ContainsKey(target - num) && dict.[target - num] <> index)
    |> Option.map (fun (num, index) -> [|index; dict.[target - num]|])
    |> Option.get

let twoSum array target = v1 array target

let print array target = 
    let result = twoSum array target
    printfn "[%d, %d]" result.[0] result.[1]

[<EntryPoint>]
let main argv =
    print [| 2; 7; 11; 15 |] 9
    print [| 3; 2; 4 |] 6
    print [| 3; 3 |] 6
    0
